In [1]:
# Misc and Fundamentals
import pandas as pd
import matplotlib.pyplot as plt
import os
import joblib

# Setup 
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.compose import make_column_transformer
from sklearn.compose import make_column_selector
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV

# Models
from sklearn import tree
from sklearn import ensemble

# Evaluation
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error

In [2]:
tv_df = pd.read_csv("../Data/data/streaming_titles_clean.csv")
# Convert columns to dummies
print(tv_df["rating"].unique())

tv_df.head(3)

[nan 'PG-13' 'R' 'TV-14' 'PG' 'TV-PG' 'NOT RATED' 'TV-MA' 'G' 'TV-G'
 'TV-Y7' 'TV-Y' 'NR' '13+' '18+' '16+' 'TV-NR' 'ALL' '7+' 'NC-17'
 'UNRATED' 'AGES_16_' 'NOT_RATE' 'TV-Y7-FV' 'UR']


type         title director cast country  release_year rating  duration  \
0  Movie  Silent Night      NaN  NaN     NaN          2020    NaN      94.0   
1  Movie  The Marksman      NaN  NaN     NaN          2021  PG-13     108.0   
2  Movie          Gaia      NaN  NaN     NaN          2021      R      97.0   

                                         description  score  ...  \
0  Mark, a low end South London hitman recently r...   56.0  ...   
1  A hardened Arizona rancher tries to protect an...   57.0  ...   
2  A forest ranger and two survivalists with a cu...   63.0  ...   

  genre.Coming_of_Age  genre.Anthology  genre.Buddy  genre.Parody  \
0               False            False        False         False   
1               False            False        False         False   
2               False            False        False         False   

   genre.Spy/Espionage  genre.Survival  genre.Soap_Opera_/_Melodrama  \
0                False           False                         False   
1                False           False                         False   
2                False           False                         False   

   genre.Dance  genre.Medical  genre.Disaster  
0        False          False           False  
1        False          False           False  
2        False          False           False  

[3 rows x 84 columns]

In [3]:
# Train/Test split
tv_train, tv_test = train_test_split(tv_df, random_state = 2023) # random state important for evaluation

X = tv_train.loc[:,tv_train.columns != "score"]
y = tv_train["score"]

The following function is to be able to save and load models without having to re-compute everything every time.

In [4]:
def fitmodel(model, filename, df = tv_train):
    
    if not os.path.isfile(filename):

        model.fit(X,y)

        joblib.dump(model, filename)

    else:
        modeltemp = joblib.load(filename)
        if (type(model) != type(modeltemp)) or \
            (tuple([k[0] for k in model.steps]) != tuple([k[0] for k in modeltemp.steps])):
            print ("\033[93m Warning: model mismatch. Delete the file {filename} and rerun or risk faulty models.\n \033[0m".format(filename=filename))
        model = modeltemp
    
    return model

In [5]:
from sklearn.metrics import get_scorer_names
# get_scorer_names()[20:]

We are going to use the `neg_root_mean_squared_error` metric, because it essentially gives us the weighted distance our model is from the correct metric.

In [6]:
from numpy import mean, std
print("mean: %0.1f, rMSE: %0.4f"% (mean(y.values),std(y.values)))

mean: 48.4, rMSE: 21.7000


A rMSE of 21.7 here means that if we guessed the mean every time, we would be on average off by 21.7 points. So, our model should try to get a better rMSE than 21.7.

In [7]:
from ModelHelpers import *

Let's start by making a Beta Regression predictor.

In [13]:
beta_predictors = columnstartswith("genre",df=tv_df) + ["rating"]

beta_pipe = Pipeline(steps =  [
    ("predictors", ColumnSelector(beta_predictors)),
    ("onehot",make_column_transformer((OneHotEncoder(),["rating"]))),
    ("beta", BetaRegression(from_range=(0,100)))
    ])

beta_grid = dict(beta__l1_ratio = [0.0,0.25,0.5,0.75,1.0],
                 beta__scale = [0.1,1,10,100],
                 beta__alpha = [100,10,1,0.1,0.01]
                 )

tv_folded = KFold(n_splits = 5).split(X,y)

beta_grid_search = GridSearchCV(estimator = beta_pipe,
                                param_grid = beta_grid,
                                cv = tv_folded,
                                scoring = 'neg_root_mean_squared_error',
                                verbose = 0)

In [14]:
beta_grid_result = beta_grid_search.fit(X,y)

c:\Users\Colton\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:592: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1225.5380692987942, tolerance: 0.24512307187618873
  model = cd_fast.sparse_enet_coordinate_descent(
c:\Users\Colton\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\model_selection\_validation.py:778: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "c:\Users\Colton\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\model_selection\_validation.py", line 767, in _score
    scores = scorer(estimator, X_test, y_test)
  File "c:\Users\Colton\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\metrics\_scorer.py", line 234, in __call__
    return self._score(
  File "c:\Users\Colton\AppData\Local\Programs\

In [10]:
print("Best parameters: {}\nBest root MSE: %0.4f.".format(beta_grid_result.best_params_) % -beta_grid_result.best_score_)
beta_pipe.set_params(**beta_grid_result.best_params_)

Best parameters: {'beta__alpha': 100, 'beta__l1_ratio': 0.0, 'beta__scale': 0.1}
Best root MSE: nan.


Pipeline(steps=[('predictors',
                 ColumnSelector(columns=['genre.Crime', 'genre.Drama',
                                         'genre.Thriller', 'genre.Action',
                                         'genre.Horror',
                                         'genre.Science_Fiction', 'genre.Music',
                                         'genre.Reality', 'genre.Romance',
                                         'genre.Comedy', 'genre.Mystery',
                                         'genre.Documentary', 'genre.History',
                                         'genre.Teen',
                                         'genre.Health_&_Wellness',
                                         'genre.Lifestyle', 'genre.Culture',
                                         'genre.Blac...
                                         'genre.Latino', 'genre.Adventure',
                                         'genre.Anime', 'genre.Talk_Show',
                                         'genre.Sketch_Comedy', 'genre.Family',
                                         'genre.Kids', 'genre.Classics',
                                         'genre.LGBTQ', 'genre.Adult_Animation',
                                         'genre.Sitcom', ...])),
                ('onehot',
                 ColumnTransformer(transformers=[('onehotencoder',
                                                  OneHotEncoder(),
                                                  ['rating'])])),
                ('beta',
                 BetaRegression(alpha=100, from_range=(0, 100), l1_ratio=0.0,
                                scale=0.1))])

In [11]:
fitmodel(beta_pipe, "models/beta_regression.joblib")
pass